unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/NFTShape/Handler/ECSNFTShapeComponentHandler.cs

```diff
@@ -1,126 +1,152 @@
+                shapeFrame.gameObject.transform.SetParent(entity.gameObject.transform);
+                shapeFrame.gameObject.transform.ResetLocalTRS();
+                renderersComponent.AddRenderer(scene, entity, shapeFrame.frameRenderer);
+            }
 
-            if (info == null)
+            Color3 modelColor = model.GetColor();
+            if (prevModel == null || !modelColor.Equals(prevModel.GetColor()))
             {
-                LoadFailed();
-                return;
+                shapeFrame.UpdateBackgroundColor(new Color(modelColor.R, modelColor.G, modelColor.B, 1));
             }
 
-            INFTAsset nftAsset = await assetRetriever.LoadNFTAsset(info.previewImageUrl);
-            
-            if (nftAsset == null)
+            if (!string.IsNullOrEmpty(model.Src) && nftLoadedScr != model.Src)
             {
-                LoadFailed();
-                return;
+                if (!infoRetrieverDisposed)
+                {
+                    infoRetriever.Dispose();
+                    infoRetrieverDisposed = true;
+                }
+                if (!assetRetrieverDisposed)
+                {
+                    assetRetriever.Dispose();
+                    assetRetrieverDisposed = true;
+                }
+                LoadNFT(model.Src);
             }
 
-            shapeFrame.SetImage(info.name, info.imageUrl, nftAsset);
-        }
-        
-        private void CreateNFTShapeFrame(IDCLEntity entity,PBNFTShape model)
-        {
-            if (shapeFrame != null)
-                DisposeShapeFrame(entity);
-            
-            shapeFrame = factory.InstantiateLoaderController(model.Style);
-           
-            entity.meshesInfo.meshRootGameObject = shapeFrame.gameObject;
-            entity.meshesInfo.currentShape = shapeFrame.shape;
-            
-            entity.meshesInfo.meshRootGameObject.name = "NFT mesh";
-            entity.meshesInfo.meshRootGameObject.transform.SetParent(entity.gameObject.transform);
-            entity.meshesInfo.meshRootGameObject.transform.ResetLocalTRS();
-            
-            dataStore.AddShapeReady(entity.entityId, entity.meshesInfo.meshRootGameObject);
+            prevModel = model;
         }
 
-        private void LoadFailed()
+        private static void DisposeShapeFrame(IParcelScene scene, IDCLEntity entity,
+            INFTShapeFrame nftFrame, IInternalECSComponent<InternalRenderers> renderersComponent)
         {
-            factory.InstantiateErrorFeedback(shapeFrame.gameObject);
-            shapeFrame.FailLoading();
+            if (nftFrame == null)
+                return;
+
+            renderersComponent.RemoveRenderer(scene, entity, nftFrame.frameRenderer);
+            nftFrame.Dispose();
+            Object.Destroy(nftFrame.gameObject);
         }
-        
-        internal void ApplyModel(PBNFTShape model)
+
+        internal async void LoadNFT(string scr)
```

Comment:
We need cancellation here.

I would have a CancellationToken external in the param and another one internal (for dispossing purposes). There are a bunch of examples in  the project using linked cancellation token in case you need a working example.

Reply:
Nevermind, I saw that the `LoadNFT` method already has cancellation inside and also de `Disposal` should cover all the cases. Disregard my comment pls!